### ISSUE TITLE
Reduce size of png images

### ISSUE THREAD

**Issue by anutator** [BookStackApp/BookStack#2342]
**Describe the feature you'd like**
Add option to reduce all loaded png images from 24-bit to 256 or even 64 colors.
https://about.gitlab.com/blog/2020/01/30/simple-trick-for-smaller-screenshots/ --- article
Program to use - pngquant
Example:
pngquant 64 --skip-if-larger --strip --ext=.png --force * 
* - all files in directory. Or you can use file names.

**Describe the benefits this feature would bring to BookStack users**
Screenshots by default save pictures as use 24-bit png files. With this feature use can reduce BookStack's folder size and open site pages quickly.

**Comment by tiredofit** [comment]
I have built this functionality using optipng and jpegoptim into a Docker image and execute it on a timer for the folders. The image includes functionality/options that are not yet supported in Bookstack official release so it may not be for you, however you are free to give it a (try)[https://github.com/tiredofit/docker-bookstack]

**Comment by tipuraneo** [comment]
Hi out there,

I wrote the following bash script to optimize uploaded images:

```
# !/bin/bash
# set -x
#
# Recompresses uploaded images of the current month and strips meta tags.
# Images are only processed if there are potential savings. Script can be run
# multiple times without recompressing already compressed images - processed
# images are skipped.Needs pngqunat and jpegoptimum to be installed. Run script
# atleast once a week better daily.
# Logs at /var/log/[script-name].log
# Metadata of png files in drawio are not stripped because they are containing
# vector data which are neccessary to edit these drawings
#
# Version 2023-09-06

FOLDER_BASE=/var/www/bookstack/public/uploads/images

LOG=/var/log/`echo $(basename "$0") | rev | cut -d. -f2- | rev`.log
#FOLDERS="cover_book cover_bookshelf drawio gallery system user"
echo "$(date '+%Y-%m-%d %H:%M:%S') === START ===" >> $LOG
FOLDERS="`ls -q $FOLDER_BASE`"
for FOLDER in $FOLDERS; do
  FOLDER_LONG=$FOLDER_BASE/$FOLDER/`date +"%Y-%m"`/
# FOLDER_LONG=$FOLDER_BASE/$i/2023-07/
  if [ -d "$FOLDER_LONG" ]; then
    DU_BEFORE=`du -s --block-size=1 $FOLDER_LONG | cut -f1`
    DU_BEFORE_K=$(($DU_BEFORE/1024))
    FILES=`ls  -lq $FOLDER_LONG | grep -iE ".*\.(jpg|jpeg|png)$" | wc -l`
    if [ $FILES -gt 0 ]; then
      if [ "$FOLDER" = "drawio" ]; then
        find $FOLDER_LONG -name "*.png" -exec pngquant --force --skip-if-larger --speed=1 --ext .png --ordered --quality=75-90 {} \;
        find $FOLDER_LONG -name "*.PNG" -exec pngquant --force --skip-if-larger --speed=1 --ext .PNG --ordered --quality=75-90 {} \;
      else
        find $FOLDER_LONG -name "*.png" -exec pngquant --strip --force --skip-if-larger --speed=1 --ext .png --ordered --quality=75-90 {} \;
        find $FOLDER_LONG -name "*.PNG" -exec pngquant --strip --force --skip-if-larger --speed=1 --ext .PNG --ordered --quality=75-90 {} \;
      fi
      find $FOLDER_LONG -iname "*.png" -exec chown -f --reference=$FOLDER_LONG {} \;
      find $FOLDER_LONG -iregex ".*\.jpe?g" -exec jpegoptim -q -p --all-progressive -m75 --strip-com --strip-exif --strip-iptc --strip-icc --strip-xmp {} \;
      DU_AFTER=`du -s --block-size=1 $FOLDER_LONG | cut -f1`
      DU_AFTER_K=$((DU_AFTER/1024))
      if [ $DU_AFTER -lt $DU_BEFORE ]; then
        SAVED_K=$((DU_BEFORE_K-$DU_AFTER_K))
        echo "$(date '+%Y-%m-%d %H:%M:%S') $FOLDER_LONG $FILES images, folder shrinked from ${DU_BEFORE_K}k to ${DU_AFTER_K}k, ${SAVED_K}k saved" >> $LOG
      else
        echo "$(date '+%Y-%m-%d %H:%M:%S') $FOLDER_LONG nothhing to do, $FILES images, ${DU_AFTER_K}k" >> $LOG
     fi
    else
      echo "$(date '+%Y-%m-%d %H:%M:%S') $FOLDER_LONG no images found" >> $LOG
    fi
  else
    echo "$(date '+%Y-%m-%d %H:%M:%S') $FOLDER_LONG does not exists" >> $LOG
  fi
done
echo "$(date '+%Y-%m-%d %H:%M:%S') $FOLDER_BASE/ `find $FOLDER_BASE/. -type f | wc -l` files, `du -sh $FOLDER_BASE | cut -f1`" >> $LOG
echo "$(date '+%Y-%m-%d %H:%M:%S') === STOP ===" >> $LOG
```

I'll maintain the script here:
https://kb.el.uni-leipzig.de/books/bookstack/page/optimierung-von-bilddateien

**Comment by PC-SPEZIALIST** [comment]
We came here after reading the [blog post about the 25.05 release](https://www.bookstackapp.com/blog/bookstack-release-v25-05/) which mentions the [support of AVIF](https://github.com/BookStackApp/BookStack/issues/5474) and links to the [amazing article about AVIF](https://jakearchibald.com/2020/avif-has-landed/) by @jakearchibald.

It would be great to have an optional setting inside [BookStack](https://www.bookstackapp.com/) to optimise uploads because I think a lot of instances would benefit massively from compression.

**Comment by ezzra** [comment]
a lot of our photos in the `scaled-1680-`  folder are still over 2 or 3 MB large, they could need a lot of compression an I really would like if bookstack could do that work automatically

**Comment by ssddanbrown** [comment]
Just to confirm some points on the current system when resizing is done:

- jpeg: are currently created at 75% quality (which generally should produce a fair balance of size/quality).
- png: Are created with default zlib compression, which should be a balance. We don't have direct control over this via the library we use, so would have to go around it.
  - There are options for indexing and interlacing, but those are probably not things I'd want to force on users as they can be scenario dependant. Could have options for it but I'm not keen on adding/maintaining options that granular.  
- webp: same a jpeg.
- avif: same as jpeg.

AVIF support is getting good now, so maybe we could convert to avif or webp.
I'd really need the default PDF renderer to support the format though. I think it already supports webp, and [I've contributed AVIF support here](https://github.com/dompdf/dompdf/pull/3703), which is awaiting a future release of the library.

It would help if examples could be shared of the larger file edge-cases (after being resized), just to double check what would help most for these, or to check if there's potentially something else going on we might have missed.

**Comment by tipuraneo** [comment]
I agree — the currently selected image formats and compression settings seem like a reasonable default.

I work a lot with both photographic images and graphics, and there are quite a few comparisons available online (for example [here](https://theimagecdn.com/docs/webp-vs-avif-vs-jpeg) or [here](https://imagerobo.com/blogs/image-formats-comparison-jpeg-webp-avif)). Roughly speaking, AVIF can reduce file size by around 50% compared to JPEG. For graphics (PNG), the reduction can sometimes be even more significant (around one third or more, depending on the image).

In my own manual workflows, I generally switch to AVIF wherever possible due to the better compression efficiency, and would [typically recommend it](https://kb.el.uni-leipzig.de/books/tipps-tricks-und-kniffe/page/bilder-webtauglich-machen#bkmrk-dateiformat) for similar use cases — especially when storage size or bandwidth matters.

To provide some concrete examples, I’ve [published a few test cases](https://kb.el.uni-leipzig.de/books/bookstack/page/spielwiese#bkmrk-avif-bild). These examples are already resized images, so they should be relevant to the use-case discussed here.

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/2342
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
